ydb_sqlalchemy/sqlalchemy/__init__.py

```diff
@@ -210,6 +210,54 @@ def visit_regexp_match_op_binary(self, binary, operator, **kw):
     def visit_not_regexp_match_op_binary(self, binary, operator, **kw):
         return self._generate_generic_binary(binary, " NOT REGEXP ", **kw)
 
+    def _is_optional(self, bind_name: str) -> bool:
+        bind = self.binds[bind_name]
+        if isinstance(bind.type, sa.Boolean):
+            return True
+        if bind_name in self.column_keys and hasattr(self.compile_state, "dml_table"):
+            if bind_name in self.compile_state.dml_table.c:
+                column = self.compile_state.dml_table.c[bind_name]
+                return not column.primary_key
+        return False
+
+    def _get_bind_type(self, bind_name: str, bind: sa.BindParameter, post_compile_bind_values: list) -> Optional[str]:
+        is_optional = self._is_optional(bind_name) or None in post_compile_bind_values
+        if not bind.expanding:
+            if isinstance(bind.type, sa.types.NullType):
+                return None
+            bind_type = bind.type.compile(self.dialect)
+        else:
+            not_null_values = [v for v in post_compile_bind_values if v is not None]
+            if not_null_values:
+                bind_type = sa.BindParameter("", not_null_values[0]).type.compile(self.dialect)
+            else:
+                return None
+        if is_optional:
+            bind_type = f"Optional<{bind_type}>"
+        return bind_type
+
+    def render_declare(self, post_compile_parameters: Union[list[dict], dict]) -> str:
+        if isinstance(post_compile_parameters, list):
+            common_keys = set.intersection(*map(set, post_compile_parameters))
```

Comment:
It is difficult to read and understand many operations in one line.
Better to split code for simpler steps - for easy read in future.

```python
parameter_names_sets = [set(parameters_dict.keys()) for parameters_dict in post_compile_parameters]
common_keys = set.intersection(parameter_names_sets)

post_compile_parameters_new = {}
for k in common_keys:
    values = [dic[k] for dic in post_compile_parameters]
    post_compile_parameters_new[k] = values
post_compile_parameters = post_compile_parameters_new

```

Reply:
Tried to make code easier to read